src/sdk/pynni/nni/pbt_tuner/pbt_tuner.py

```diff
@@ -34,19 +38,102 @@ def exploit_and_explore(bot_trial_info, top_trial_info, factors, epoch, search_s
     bot_checkpoint_dir = bot_trial_info.checkpoint_dir
     top_hyper_parameters = top_trial_info.hyper_parameters
     hyper_parameters = copy.deepcopy(top_hyper_parameters)
-    # TODO think about different type of hyperparameters for 1.perturbation 2.within search space
+    random_state = np.random.RandomState()
     for key in hyper_parameters.keys():
         if key == 'load_checkpoint_dir':
             hyper_parameters[key] = hyper_parameters['save_checkpoint_dir']
         elif key == 'save_checkpoint_dir':
             hyper_parameters[key] = os.path.join(bot_checkpoint_dir, str(epoch))
-        elif isinstance(hyper_parameters[key], float):
-            perturb = np.random.choice(factors)
-            val = hyper_parameters[key] * perturb
+        elif search_space[key]["_type"] == "choice":
+            choices = search_space[key]["_value"]
+            can_sort = True
+            for item in choices:
+                if isinstance(item, str):
+                    can_sort = False
+            if can_sort:
+                choices.sort()
+            if random.random() < resample_probability or hyper_parameters[key] not in choices:
+                hyper_parameters[key] = parameter_expressions.choice(choices, random_state)
+            elif random.random() > 0.5:
+                hyper_parameters[key] = choices[max(0, choices.index(hyper_parameters[key]) - 1)]
+            else:
+                hyper_parameters[key] = choices[min(len(choices) - 1, choices.index(hyper_parameters[key]) + 1)]
+        elif search_space[key]["_type"] == "randint":
             lb, ub = search_space[key]["_value"][:2]
-            if search_space[key]["_type"] in ("uniform", "normal"):
-                val = np.clip(val, lb, ub).item()
-                hyper_parameters[key] = val
+            if random.random() < resample_probability:
+                hyper_parameters[key] = parameter_expressions.randint(lb, ub, random_state)
+            elif random.random() > 0.5:
+                hyper_parameters[key] = min(hyper_parameters[key] + 1, ub)
+            else:
+                hyper_parameters[key] = max(hyper_parameters[key] - 1, lb)
```

Comment:
these lines are almost the same as other search space types, please use a function instead.

Reply:
+1

Reply:
modified